src/anomalib/utils/callbacks/thresholding.py

```diff
@@ -0,0 +1,159 @@
+
+    def on_validation_epoch_start(self, trainer: "trainer.AnomalibTrainer", pl_module: AnomalyModule) -> None:
+        self._reset(pl_module)
+
+    def on_validation_batch_end(
+        self,
+        trainer: "trainer.AnomalibTrainer",
+        pl_module: AnomalyModule,
+        outputs: STEP_OUTPUT | None,
+        batch: Any,
+        batch_idx: int,
+        dataloader_idx: int = 0,
+    ) -> None:
+        if outputs is not None:
+            self._outputs_to_cpu(outputs)
+            self._update(pl_module, outputs)
+
+    def on_validation_epoch_end(self, trainer: "trainer.AnomalibTrainer", pl_module: AnomalyModule) -> None:
+        self._compute(pl_module)
+
+    def _setup(self, pl_module: AnomalyModule) -> None:
+        pl_module.image_threshold = self.image_threshold
+        pl_module.pixel_threshold = self.pixel_threshold
+
+    def _update(self, pl_module: AnomalyModule, outputs: STEP_OUTPUT) -> None:
+        pl_module.image_threshold.cpu()
+        pl_module.image_threshold.update(outputs["pred_scores"], outputs["label"].int())
+        if "mask" in outputs.keys() and "anomaly_maps" in outputs.keys():
+            pl_module.pixel_threshold.cpu()
+            pl_module.pixel_threshold.update(outputs["anomaly_maps"], outputs["mask"].int())
+
+    def _reset(self, pl_module: AnomalyModule) -> None:
+        pl_module.image_threshold.reset()
+        pl_module.pixel_threshold.reset()
+
+    def _compute(self, pl_module: AnomalyModule) -> None:
+        pl_module.image_threshold.compute()
+        if pl_module.pixel_threshold._update_called:
+            pl_module.pixel_threshold.compute()
+        else:
+            pl_module.pixel_threshold.value = pl_module.image_threshold.value
+
+    def _outputs_to_cpu(self, output: STEP_OUTPUT):
+        if isinstance(output, dict):
+            for key, value in output.items():
+                output[key] = self._outputs_to_cpu(value)
+        elif isinstance(output, Tensor):
+            output = output.cpu()
+        return output
+
+    def _initialize_thresholds(
+        self,
+        threshold: BaseAnomalyThreshold
+        | tuple[BaseAnomalyThreshold, BaseAnomalyThreshold]
+        | DictConfig
+        | ListConfig
+        | str,
+    ) -> None:
+        if isinstance(threshold, BaseAnomalyThreshold):
+            self.image_threshold = threshold
+            self.pixel_threshold = threshold
+        elif isinstance(threshold, tuple):
+            self.image_threshold = threshold[0]
+            self.pixel_threshold = threshold[1]
+        elif isinstance(threshold, (str, DictConfig, ListConfig)):
+            self._load_from_config(threshold)
+        else:
+            raise ValueError(f"Invalid threshold type {type(threshold)}")
+
+    def _load_from_config(self, threshold: DictConfig | str | ListConfig) -> None:
+        """Loads the thresholding class based on the config.
+
+        Example:
+            threshold: F1AdaptiveThreshold
```

Comment:
is it possible to do:

```yaml
threshold: adaptive
```

Reply:
This refactor cannot handle adaptive. I can check against the lowercased substring but it might become ambigious if another threshold is added with adaptive in it

Reply:
Yeah that's fair. I just thought about backward compatibility perspective, but we're going to break it anyway